pkg/controller/notification/notification_controller.go

```diff
@@ -0,0 +1,143 @@
+)
+
+var log = logf.Log.WithName("controller_notification")
+
+// Add creates a new Notification Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, config *configuration.Config) error {
+	return add(mgr, newReconciler(mgr, config))
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("notification-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource Notification
+	err = c.Watch(&source.Kind{Type: &toolchainv1alpha1.Notification{}},
+		&handler.EnqueueRequestForObject{},
+		predicate.GenerationChangedPredicate{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager, config *configuration.Config) reconcile.Reconciler {
+	return &ReconcileNotification{client: mgr.GetClient(), scheme: mgr.GetScheme(), config: config}
+}
+
+var _ reconcile.Reconciler = &ReconcileNotification{}
+
+// ReconcileNotification reconciles a Notification object
+type ReconcileNotification struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+	config *configuration.Config
+}
+
+func (r *ReconcileNotification) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling Notification")
+
+	// Fetch the Notification instance
+	notification := &toolchainv1alpha1.Notification{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, notification)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// if is delivered, then check when status was changed and delete it if the requested duration has passed
+	completeCond, found := condition.FindConditionByType(notification.Status.Conditions, toolchainv1alpha1.NotificationSent)
+	if found && completeCond.Status == corev1.ConditionTrue {
+		deleted, requeueAfter, err := r.checkTransitionTimeAndDelete(reqLogger, notification, completeCond)
+		if deleted || err != nil {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{
+			Requeue:      true,
+			RequeueAfter: requeueAfter,
+		}, nil
+	}
+
+	reqLogger.Info("Notification has been delivered")
```

Comment:
I would slightly refactor that code.

1. Log "Notification has been sent"
2. Set the status to Sent to true.
3. Then if step 2 returns an error, log the error and return it.
4. Requeue with DurationBeforeNotificationDeletion.

Otherwise you don't log the error if there is any errors while updating the status.

Reply:
Sure, this will need to be fixed for both ChangeTierRequest and Notification